カレンダーのプログラム(ruby)の課題提出

88-99 opened the pull request:
カレンダーのプログラム(ruby)の課題を実装しました。
ご確認お願いいたします。

Commit:
first commit

func09 commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,213 @@
+
+argv = opt.parse(ARGV)
+
+if argv[0].to_i == 0
+  @month = Date.today.month
+else
+  @month = argv[0].to_i
+end   
+
+if argv[1].to_i == 0
+  @year = Date.today.year
+else
+  @year = argv[1].to_i
+end
```
```
ruby calendar.rb -y 2000 -m 1
```

このような順番でコマンドを実行すると

```
puts "#{@year}年#{@month}月" #=> 1年2000月
```

となってしまって、データがおかしくなるようです。修正してください〜

88-99 replied:
配列のインデックスを指定しているので、引数の順番が入れ替わるとおかしくなってしまいますね。考えます。

88-99 replied:
ARGV.getoptsを使用して修正しました。-mと-yを入れ替えてもセットした引数に値が引き継がれます。

func09 replied:
確認しましたー、良いと思います！

func09 commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,213 @@
+  @year = argv[1].to_i
+end
+
+def check_first_wday(y = @year, m = @month)
+  @first_wday = Date.new(y, m, 1).wday
+end
+check_first_wday
```
ここで check_first_wday のメソッドを定義する理由はどうしてでしょうか？

* 定義してすぐ実行している
* メソッドの内容が１行
* 定義したメソッドを１回しか使っていない

などの理由から、わざわざメソッドを定義する必要はなさそうで

```suggestion
@first_wday = Date.new(@year, @month, 1).wday
```

これだけでよさそうに見えました。

88-99 replied:
ごもっともです。メソッドの書き方を覚えたので使ってみたいという気持ちがあり、メソッドにしなくても動くということに気づいていませんでした。end_dateも類似しているので、first_wdayとend_dateの部分を修正いたします。

88-99 replied:
メソッドの定義をやめ、インスタンス変数の＠を削除、ローカル変数に変更しました。

func09 replied:
👍

func09 commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,213 @@
+check_end_date
+
+puts "      #{@month}月 #{@year}"
+print ["日", "月", "火", "水", "木", "金", "土\n"] * (" ")
```
```suggestion
puts ["日", "月", "火", "水", "木", "金", "土"].join(" ")
```

こんな感じでも良いと思います。

88-99 replied:
修正します。

88-99 replied:
修正しました。

func09 commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,213 @@
+puts "      #{@month}月 #{@year}"
+print ["日", "月", "火", "水", "木", "金", "土\n"] * (" ")
+
+def week1
+  case @first_wday
+  when 0
+    print " 1  2  3  4  5  6  7\n"
+  when 1
+    print "    1  2  3  4  5  6\n"
+  when 2
+    print "       1  2  3  4  5\n"
+  when 3
+    print "          1  2  3  4\n"
+  when 4
+    print "             1  2  3\n"
+  when 5
+    print "                1  2\n"
+  when 6
+    print "                   1\n"
+  end
+end
```
week1 ~ week6 のメソッドに関しては、あらゆる組み合わせを想定して出力パターンを用意していますね。

これだとカレンダー程度の組み合わせならなんとかなりますが、
組み合わせが多くなるとコードが膨大な量になりますし、バグ修正なども大変になるので
もう少しプログラミングが得意な計算を使った解決方法を考えてみましょう。

func09 replied:
例えば（これは正解ではなくてヒントですが）
まず問題を小さく分解して考えてみましょう。

* 表示する日付をすべて取得して、配列にする
* 表示する日付をひとつずつ出力して、土曜日の後で改行する
* 表示を整える

## 表示する日付をすべて取得して、配列にする

```ruby
first_date = Date.new(@year, @month, 1)
end_date = Date.new(@year, @month, -1)
days_in_month = (first_date .. end_date).to_a # Rangeクラスを作成した後に配列にする
```

## 表示する日付をひとつずつ出力して、土曜日の後で改行する

```ruby
days_in_month.each do |date|
  print "#{date.day} "
  print "\n" if date.saturday? # 土曜日の時は改行する
end
```

これで実行すると、下記のような表示になります。

```
1 2 3 4 5 6
7 8 9 10 11 12 13
14 15 16 17 18 19 20
21 22 23 24 25 26 27
28 29 30 31
```

問題の解き方はひとつではありませんが、この方法では `1` と `-1` 以外の具体的な数字を使うことなく
カレンダーっぽい表示ができています。

こんな感じの考え方をベースにコードを組み立ててみてください。

88-99 replied:
詳細なご説明とヒントありがとうございます。日にちのベタ貼りは気になっている点でした。恐れ入ります。

88-99 replied:
修正しました。case文とfirst_dateの前のスペースを入れる部分について短縮できないものかと思い考えてみましたが、解決策を思い付きませんでした。ご確認お願いいたします。

func09 reviewed (changes requested):
レビューしました。
日付部分をベタ書きする解決策は、
プログラミング的な解決策をじっくり考えてみましょう。

Commit:
コマンドから引数を取得する部分の修正。メソッドと曜日表示の書き方変更。

変更前のコードでは、引数を入れ替えるとおかしくなってしまうため。メソッドの書き方については、メソッドを利用するメリットがないので、メソッドを使わない書き方に変更。

Commit:
days_in_manth.each do |date|〜endの中身を修正。ファイル全体の@を削除。

コードを短くして日にちを揃えました。@はなくても機能するため削除しました。

func09 commented on 02.calendar/calendar.rb:
```diff
@@ -0,0 +1,51 @@
+
+days_in_manth = (first_date..end_date).to_a
+
+days_in_manth.each do |date|
+  if date.day < 10
+    case
+    when first_wday == 1 && date.day == 1
+      print "    #{first_date.day.to_s} "
+    when first_wday == 2 && date.day == 1
+      print "       #{first_date.day.to_s} "
+    when first_wday == 3 && date.day == 1
+      print "          #{first_date.day.to_s} "
+    when first_wday == 4 && date.day == 1
+      print "             #{first_date.day.to_s} "
+    when first_wday == 5 && date.day == 1
+      print "                #{first_date.day.to_s} "
+    when first_wday == 6 && date.day == 1
+      print "                   #{first_date.day.to_s} "
+    else
+      print " #{date.day} "
+    end
+  else
+    print "#{date.day} "
+  end
+  print "\n" if date.saturday? || date == end_date
+end
```
まず現状のコードでも提出物としては十分許容範囲内です。
その上で、コードをスマートにするサンプルを書いてみますね。

# １日の左側のスペースを分岐なしで出力するサンプル

```suggestion
print "   " * first_wday # スペース３個を first_day で掛けた分スペースを入れる
days_in_manth.each do |date|
  if date.day < 10
    print " #{date.day} "
  else
    print "#{date.day} "
  end
  print "\n" if date.saturday? || date == end_date
end
```

１日を出力しはじめるスペースの数は、上のような計算で出すことができます。

* 日曜日は `0` なので、左側のスペースは `0`
* 月曜日は `1` なので、左側のスペースは `3`
* 火曜日は `2` なので、左側のスペースは `6`

と考えていくと、どうやら

```
スペース3個 * first_wday
```

だということがわかります。

func09 replied:
上のコードをもう一歩スマートにするために

* 日付が１桁でも２桁、同じコードで出力する

という課題を解いてみます。

# 日付が１桁でも２桁でも分岐しないで文字幅を揃えるサンプル

```suggestion
print "   " * first_wday # スペース３個を first_day で掛けた分スペースを入れる
days_in_manth.each do |date|
  print date.day.to_s.rjust(2) + " "
  print "\n" if date.saturday? || date == end_date
end
```

`String#rjust` https://docs.ruby-lang.org/ja/latest/method/String/i/rjust.html という文字列の幅を揃えるメソッドを使います。

適切にコードが短くなると見通しがよくなって理解しやすくなったり、バグが入りづらくなることにつながるので意識してみてください。

88-99 replied:
> 適切にコードが短くなると見通しがよくなって理解しやすくなったり、バグが入りづらくなることにつながるので意識してみてください。

承知いたしました。ありがとうございます。
[ 理解のまとめ ]
- `date.day`には`days_in_month = (first_date..end_date).to_a`によって`first_date`が入るので、`&& date.day == 1`は省略可能。
- `print`は改行コードを入れなければ改行されないため、この方法で最初にスペースを挟むことができる。
- 1日は一桁表示用の半角を抜いたら、半角スペース3個が一つの単位になっていて、`first_wday`が1増えるごとに、半角スペースが3増える法則になっている。よって`print "   " * first_wday`と表すことができる。
`first_wday == 1 `のとき半角スペースは3個
`first_wday == 2`のとき半角スペースは6個
・
・
`first_wday == 6`のとき半角スペースは18個

- rjust(2)することで、1桁数字のときは半角スペースを一つ入れて数字を右に寄せることができる。

func09 reviewed (approved):
コメントしましたが、現状でも十分なのでLGTMとしました〜！
